rehttp_server_post17_test.go

```diff
@@ -0,0 +1,136 @@
+//go:build go1.7
+// +build go1.7
+
+package rehttp
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestTransport_RoundTripTimeouts(t *testing.T) {
+	attemptCountMu := sync.Mutex{}
+	attemptCount := 0
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		attemptCountMu.Lock()
+		attemptCount++
+		attemptCountMu.Unlock()
+		time.Sleep(time.Millisecond * 100)
+		w.WriteHeader(http.StatusTooManyRequests)
+	}))
+	defer ts.Close()
+
+	tr := NewTransport(
+		http.DefaultTransport,
+		RetryAll(RetryMaxRetries(3), RetryAny(
+			RetryStatuses(http.StatusTooManyRequests), // retry 429s
+			func(attempt Attempt) bool { // retry context deadline exceeded errors
+				return attempt.Error != nil && attempt.Error == context.DeadlineExceeded // errors.Is requires go 1.13+
+			})),
+		ConstDelay(time.Duration(0)),
+	)
+	tr.PerAttemptTimeout = time.Millisecond * 10 // short timeout
+
+	client := http.Client{
+		Transport: tr,
+	}
+
+	req, err := http.NewRequest(http.MethodGet, ts.URL, nil)
+	if err != nil {
+		t.Errorf("error creating request: %s", err)
+	}
+
+	_, err = client.Do(req)
+	time.Sleep(time.Millisecond * 100) // let the server finish sleeping
+	// should have attempted 4 times but errored due to timeout
+	attemptCountMu.Lock()
+	if attemptCount != 4 {
+		t.Errorf("attempt count does not match expected: got %d, want %d", attemptCount, 4)
+	}
+	if err == nil {
+		t.Error("expected timeout error doing request but got nil")
+	}
+
+	attemptCount = 0
+	attemptCountMu.Unlock()
+
+	// now increase the timeout restriction
+	tr.PerAttemptTimeout = time.Second
+	res, err := client.Do(req)
+	attemptCountMu.Lock()
+	// should have attempted 4 times without going over the timeout
+	if attemptCount != 4 {
+		t.Errorf("attempt count does not match expected: got %d, want %d", attemptCount, 4)
+	}
+	if err != nil {
+		t.Errorf("got unexpected error doing request: %s", err)
+	}
+	if res == nil || res.StatusCode != http.StatusTooManyRequests {
+		t.Errorf("status code does not match expected: got %d, want %d", res.StatusCode, http.StatusTooManyRequests)
+	}
+
+	attemptCount = 0
+	attemptCountMu.Unlock()
+
+	// now remove the timeout restriction
+	tr.PerAttemptTimeout = time.Duration(0)
+	res, err = client.Do(req)
+	attemptCountMu.Lock()
+	// should have attempted 4 times without going over the timeout
+	if attemptCount != 4 {
+		t.Errorf("attempt count does not match expected: got %d, want %d", attemptCount, 4)
+	}
+	if err != nil {
+		t.Errorf("got unexpected error doing request: %s", err)
+	}
+	if res == nil || res.StatusCode != http.StatusTooManyRequests {
+		t.Errorf("status code does not match expected: got %d, want %d", res.StatusCode, http.StatusTooManyRequests)
+	}
+	attemptCountMu.Unlock()
+}
```

Comment:
What I think would be good to have is a test with both a per-attempt and an overall timeout, and a case where the overall triggers instead of the per-attempt?